### PR TITLE
Add picture display associate with a tag

### PR DESCRIPTION
**Pull request by HaominZhang**

### CODE REVIEW

File: app/assets/javascripts/tags.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (rmacklin) [inline app/assets/javascripts/tags.coffee]: I know this file came from the rails generator, but I would suggest removing it. We may add some javascript for tags, but that will come later (and it won't be coffeescript), so my preference would be to only commit files we're currently using.

File: app/assets/stylesheets/tags.scss [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the tags controller here.
+// They will automatically be included in application.css.
+// You can use Sass (SCSS) here: http://sass-lang.com/
```
Comment (rmacklin) [inline app/assets/stylesheets/tags.scss]: Same as above: I wouldn't commit this generated file if we aren't using it yet.

File: app/controllers/pictures_controller.rb [hunk 3]
```diff
@@ -2,7 +2,12 @@ class PicturesController < ApplicationController
   before_action :set_picture, only: [:show, :destroy]
 
   def index
-    @pictures = Picture.all
+    picture_collection = Picture.order('created_at DESC')
+    if params[:name]
+      @pictures = picture_collection.tagged_with(params[:name])
```
Comment (rmacklin) [inline app/controllers/pictures_controller.rb]: This is a bit confusing. `params[:name]` holds the name of a `Tag`. I would prefer if we used `params[:tag]` for this, to make it clear we are filtering `Picture`s by a `Tag`, rather than by the [`name` attribute](https://github.com/AppFolioOnboarding/Haomin_ImageUploader/blob/a92554d43ae3fef9ec453f814d8c88044bba44ea/app/models/picture.rb#L2).

File: app/helpers/tags_helper.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+module TagsHelper
+end
```
Comment (rmacklin) [inline app/helpers/tags_helper.rb]: Same as above, I wouldn't commit this generated file if we aren't using it yet.

File: app/views/layouts/_navigation.html.erb [hunk 5]
```diff
@@ -5,8 +5,8 @@
         <li class="nav-item active">
           <%= link_to 'Image', root_path, class:'nav-link'%>
         </li>
-        <li class="nav-item">
-          <a class="nav-link" href="#">Tags</a>
+        <li class="nav-item active">
```
Comment (rmacklin) [inline app/views/layouts/_navigation.html.erb]: We are misusing the `active` class here. The semantics of the `active` class are to indicate which navigation item the current page belongs to. As mentioned in https://github.com/AppFolioOnboarding/Haomin_ImageUploader/pull/11#discussion_r51194249, Aaron left it as an exercise to dynamically add the `active` class to the proper navigation item.

File: app/controllers/pictures_controller.rb (hunk 3 above)
Comment (HaominZhang) [inline app/controllers/pictures_controller.rb]: Good point! That would be more meaningful!

File: app/helpers/tags_helper.rb (hunk 4 above)
Comment (HaominZhang) [inline app/helpers/tags_helper.rb]: I'll remove them and above same issue.

File: app/views/pictures/index.html.erb [hunk 6]
```diff
@@ -1,25 +1,28 @@
...
-      <%=link_to "New Image", new_picture_path, class: 'btn btn-warning' %>
+  <p id="notice"><%= notice %></p>
+  <% unless params[:name] %>
+    <div class="row">
+      <div class="col-sm-12">
+        <%= link_to "New Image", new_picture_path, class: 'btn btn-warning', id: 'add_button' %>
+      </div>
     </div>
-  </div>
+  <% end %>
```
Comment (rmacklin) [inline app/views/pictures/index.html.erb]: Interesting. I don't necessarily think we need to hide this button if we are viewing images filtered by a tag. In fact, perhaps even better would be:
- Always render the button
- If there is a tag filter, change the button to "New #mytag image" and have the button autofill "mytag" in the tag_list on the form.

You don't have to do that, but it could be a fun exercise. In either case, I would probably not hide the New Image button even if there is a filter applied.

File: app/views/layouts/_navigation.html.erb (hunk 5 above)
Comment (HaominZhang) [inline app/views/layouts/_navigation.html.erb]: I haven't dealt with the case the we should highlight the current tab. I think it's just some copy&paste mistake I made. I will delete the `active` later and I will solve the highlight issue in future branches. Thanks!

Comment (HaominZhang) [inline app/views/layouts/_navigation.html.erb]: My staging app is up now. I forgot to run migration after last merge. Sorry I didn't notice the comment you left on my last PR.

File: app/views/pictures/show.html.erb [hunk 7]
```diff
@@ -11,7 +11,9 @@
   </p>
   <p>
     <strong>Tags:</strong>
-    <%= @picture.tag_list%>
+    <% @picture.tag_list.each do |tag| %>
+      <%= link_to tag, pictures_path(tag) %>
```
Comment (rmacklin) [inline app/views/pictures/show.html.erb]: Did you mean to use `pictures_path(name: tag)` here? If so, perhaps we are missing some test coverage that would've caught this.

File: test/controllers/pictures_controller_test.rb [hunk 8]
```diff
@@ -5,7 +5,8 @@ class PicturesControllerTest < ActionController::TestCase
   test 'can access index' do
     get :index
     assert_response :success
-    assert_select '#pictures img', 2
+    assert_select '#pictures img', 3
```
Comment (rmacklin) [inline test/controllers/pictures_controller_test.rb]: Personally I prefer when tests create the data that they use for assertions because then the only thing you need to look at is the test method in question. So, I would rather we delete the `Picture` fixtures and manually create three images in this test. Do you prefer using fixtures?

Comment (HaominZhang) [inline test/controllers/pictures_controller_test.rb]: In the beginning I think using fixtures seems more convenient but now I cannot create images completely through fixtures because there is no `tag_list` attribute in picture model. So I think I should change to the create three images in this test.

File: app/views/pictures/show.html.erb (hunk 7 above)
Comment (HaominZhang) [inline app/views/pictures/show.html.erb]: Yeah you are right. I didn't notice this part, and actually it was a mistake that I should put `tag: tag` there. I will change it.

File: test/controllers/pictures_controller_test.rb [hunk 9]
```diff
@@ -66,5 +67,25 @@ class PicturesControllerTest < ActionController::TestCase
...
+
+    get :index, name: 'tag2'
 
+    assert_response :success
+    assert_select '#pictures img', 2
+
+    assert_select 'img' do
+      assert_select ":match('src', '#{test_picture1.link}')"
+      assert_select ":match('src', '#{test_picture2.link}')"
+    end
```
Comment (rmacklin) [inline test/controllers/pictures_controller_test.rb]: You could combine assertions from lines 82-87 as:
```ruby
assert_select '#pictures img' do |elements|
  assert_equal [pic1.link, pic2.link], elements.map { |el| el[:src] }
end
```

File: app/views/layouts/_navigation.html.erb (hunk 5 above)
Comment (rmacklin) [inline app/views/layouts/_navigation.html.erb]: It's okay. Are you getting GitHub email notifications for PRs that you participate in? If not, you should head to https://github.com/settings/notifications and check the "Email" box under "Participating":
screen shot 2016-01-28 at 4 06 12 pm
Then you won't miss comments on older PRs.

File: test/controllers/pictures_controller_test.rb (hunk 9 above)
Comment (HaominZhang) [inline test/controllers/pictures_controller_test.rb]: This is really neat! I will change to this.

File: app/views/pictures/index.html.erb (hunk 6 above)
Comment (HaominZhang) [inline app/views/pictures/index.html.erb]: For now I just remove the hiding `create` link, I will try to figure out the autofill stuff in future branches.

Comment (rmacklin) [inline app/views/pictures/index.html.erb]: Yes, I think that's appropriate :+1:

File: app/helpers/application_helper.rb [hunk 10]
```diff
@@ -1,2 +1,20 @@
...
+        end
+      elsif !Picture.exists?(@picture.id)
+        current_page?(new_picture_path) ? 'active' : ''
+      else
+        current_page?(picture_path(@picture)) ? 'active': ''
+      end
+    else
+      ''
+    end
+  end
```
Comment (rmacklin) [inline app/helpers/application_helper.rb]: This logic is hard to follow. Additionally, it's breaking ruby convention because the method name ends in `?` but the method does not return a boolean. That also indicates an area where we could better separate our concerns: the helper method should be decoupled from the `active` CSS class (it could just return true or false and then the view could add the `active` class only if the method returns true).

But taking a step back, do you think this is the right abstraction to make? Could we do any better/make things clearer with an even better separation?

Comment (HaominZhang) [inline app/helpers/application_helper.rb]: First of all I agree with you that the method name should be modified and the helper method should be decoupled from CSS(I have to find a way to do this).
And for the second part, do you mean I should put this part into a future branch related to UI to make this branch much clearer?

Comment (rmacklin) [inline app/helpers/application_helper.rb]: > And for the second part, do you mean I should put this part into a future branch

Yes, I was going to suggest that. I think this "mark the correct nav item as active" challenge requires some more thought (and I'm happy to talk through it with you), so I would suggest removing it from this branch (just make it so that none of the navigation items have the `active` class for now) and tackled separately.

Comment (HaominZhang) [inline app/helpers/application_helper.rb]: Sounds good. I will remove this part from current branch. I remember we have a pair on Wednesday, maybe we can solve the problem at that time.

Comment (rmacklin) [inline app/helpers/application_helper.rb]: Sounds good :+1:

**Comment by HaominZhang** [comment]
@rmacklin I made a new commit in which I remove the navbar highlight feature. Please let me know if there is anything need to be modified or I can just merge the PR. Thank you!

File: test/controllers/pictures_controller_test.rb [hunk 11]
```diff
@@ -66,5 +79,16 @@ class PicturesControllerTest < ActionController::TestCase
     end
   end
 
+  test 'display pictures associated with a tag' do
+    pic1 = Picture.first
+    pic2 = Picture.last
```
Comment (rmacklin) [inline test/controllers/pictures_controller_test.rb]: It would be more robust to store a reference to the created `Picture`s in the `setup` block and use that reference here rather than relying on them being `Picture.first` and `Picture.last`.

File: test/controllers/pictures_controller_test.rb [hunk 12]
```diff
@@ -66,5 +79,16 @@ class PicturesControllerTest < ActionController::TestCase
...
+    get :index, name: 'tag2'
```
Comment (rmacklin) [inline test/controllers/pictures_controller_test.rb]: We should create a third picture which does not include the `tag2` tag so that we can verify that _only_ the images with `tag2` are being shown, rather than _all_ images (currently those two things are equivalent because our we did not create any images without the `tag2` tag).

File: test/controllers/pictures_controller_test.rb (hunk 11 above)
Comment (HaominZhang) [inline test/controllers/pictures_controller_test.rb]: I am not 100% percent sure if I understand your concern correctly, do you mean I should do something like `@pic1 = Picture.create!(name: xxx, link: xxx, tag_list: xxx)` so I can use `@pic1` in following test cases?

File: test/controllers/pictures_controller_test.rb (hunk 12 above)
Comment (HaominZhang) [inline test/controllers/pictures_controller_test.rb]: Good catch! I will modify this part.

**Comment by HaominZhang** [comment]
I modified that part as I mentioned and please let me know if it is correct. 
https://github.com/AppFolioOnboarding/Haomin_ImageUploader/pull/12#discussion_r51524148

File: test/controllers/pictures_controller_test.rb [hunk 13]
```diff
@@ -1,15 +1,25 @@
 require 'test_helper'
 
 class PicturesControllerTest < ActionController::TestCase
+  def setup
+    # Picture.create!([{ name: 'test_picture1', link: 'https://media.giphy.com/media/iXgxDPDb6Kcco/giphy.gif', tag_list: 'tag1, tag2' },
+    #                 { name: 'test_picture2', link: 'https://media.giphy.com/media/NS99NUGjn3pK/giphy.gif', tag_list: 'tag2, tag3' },
+    #                 { name: 'test_picture3', link: 'https://media.giphy.com/media/9Em15XLdEkfSM/giphy.gif', tag_list: 'tag5' }])
```
Comment (rmacklin) [inline test/controllers/pictures_controller_test.rb]: Can we remove this commented out part?

File: test/controllers/pictures_controller_test.rb (hunk 11 above)
Comment (rmacklin) [inline test/controllers/pictures_controller_test.rb]: Yes, precisely.

File: test/controllers/pictures_controller_test.rb (hunk 13 above)
Comment (HaominZhang) [inline test/controllers/pictures_controller_test.rb]: Sorry I forgot to remove them. I will remove them and do a commit amend

Comment (HaominZhang) [inline test/controllers/pictures_controller_test.rb]: @rmacklin fixed.

**Comment by HaominZhang** [comment]
@rmacklin I removed the comment and force pushed to this branch. Please let me know if there are other concerns. Thanks!

**Comment by rmacklin** [comment]
I think that was my only remaining comment. Have you verified the acceptance criteria on the review app?

**Comment by HaominZhang** [comment]
@rmacklin Yeah I tested on the review app and it works fine.
